init.py

```diff
@@ -180,24 +181,89 @@ def sierpinski(t, l, r, deg, preds, g):
             add_edges(l,t,preds,g)
             add_edges(t,r,preds,g)
             add_edges(r,l,preds,g)
-            
+
     def add_edges(u,v,preds,g):
         ''' Adds edges between vertices u and v for all predicates. '''
         for p in preds:
             g += [[u,p,v]]
             g += [[v,p,u]]
-            
+
     def _idgen():
         ''' Generates integer identifiers for vertices. '''
         c = 4
         while True:
             yield c
             c += 1
-            
+
     ids = _idgen()
     graph = []
-    sierpinski(1,2,3,degree,predicates, graph)   
-    with open(os.path.join(target_dir, 'sierpinskigraph_%s.txt'%(degree)), 'w') as out_file:
+    sierpinski(1,2,3,degree,predicates, graph)
+    filename = 'sierpinskigraph_%d_%s.txt' % (degree,''.join(predicates))
+    _to_file(target_dir, filename, graph)
+
+def gen_kuijpers_random(target_dir, vertices, k, labels='A B C D'):
+    '''Generate random graphs as defined on J. Kuijpers's paper "An
+    Experimental Study of Context-Free Path Query Evaluation Methods".
+    '''
+    n = vertices
+    L = labels.split(' ')
+    assert n >= k, 'It is required that vertices >= k'
+
+    def label(L):
+        '''Randomly choose a label in `L`'''
+        return random.choice(L)
+
+    def add(D,i,j,I,var):
+        '''Add an edge to `D`.'''
+        l = label(L)
+        D += [[i,l,j]]
+        D += [[j,l+"R",i]]
+        if i == len(I):
+            I += [0]
+        I[j] += 1
+        I[i] += 1
+        var['sum'] += 2
+
+    def p(i,I,sumI):
+        '''Calculate the probability of adding an edge to vertex `i`.'''
+        if sumI == 0:
+            return 1
+        return (I[i]) / sumI
+
+    D = [] # the graph
+    I = [] # vertices' degree
+    var = {'sum' : 0} # sum of vertices' degree
+    for i in range(k):
+        for j in range(k):
+            if i != j:
+                l = label(L)
+                D += [[i,l,j]]
+                D += [[j,l+"R",i]]
+    I = [(k-1)*2]*k
+    var['sum'] = (k-1)*2*k
+    for i in range(k,n):
+        chosenset = set()
+        chosensum = 0.0
+        for _ in range(k):
+            r = random.uniform(0,1-chosensum)
```

Comment:
I dont understand why we need to subtract `chosensum` from segment of uniform distribution. 
As far as I understand, when `chosensum` is equal to 0, then we select vertex j with probability pj. But then we consider segment of (0, 1-pj) (after we have chosen vertex j). Due to it we decrease probability of choice the last vertex (as far as I understand). So, could you explain, why we need to reduce segment of uniform distribution?

Reply:
In fact, this is part of the logic for choosing non-repeated values.
But, as I said below, you may reject it.

Reply:
Thanks to you, i have found a mistake in my implementation. `random.choices` return **repeated** values, so my implementation is incorrect. I will try to find the built in way to achieve right sampling.

Reply:
Nice. Let me know if you find it.

Reply:
I havent found built in function, but there is random.choice in numpy. It has flag `replace` for it.